### PR TITLE
Add tests 

### PR DESCRIPTION
**Pull request by zinagithub**

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+  validates_presence_of :name
+
+  has_many :posts, dependent: :destroy
+  has_many :friend_requests_sent, foreign_key: :sender_id,
+            class_name: "FriendRequest", dependent: :destroy
+  has_many :friend_requests_received, foreign_key: :receiver_id,
+			class_name: "FriendRequest", dependent: :destroy
+  has_many :comments, dependent: :destroy
+  has_many :friendships, dependent: :destroy
```
Comment (rvvergara) [inline app/models/user.rb]: Please see #4 for a persistent bug in creating friendships

File: test/models/user_test.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
+require 'test_helper'
+
+class UserTest < ActiveSupport::TestCase
+  def setup
+  	@user = users(:one)
+  end	
+  
+  test 'user has many friends' do
+    assert 1 < @user.friends.count
```
Comment (rvvergara) [inline test/models/user_test.rb]: Please run me through how you generated your friendships for your test user. I looked into your users fixture and didn't find any line there that you created a friendship.

File: test/models/user_test.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+class UserTest < ActiveSupport::TestCase
+  def setup
+  	@user = users(:one)
+  end	
+  
+  test 'user has many friends' do
+    assert 1 < @user.friends.count 
+  end
+
+  test 'user has many posts' do
```
Comment (rvvergara) [inline test/models/user_test.rb]: Please also run me through on how you created posts for your test users

File: test/models/friendship_test.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  end	
+  test 'Valid friendship' do
+	  assert @friendship.valid?  
+  end
+
+  test 'Invalid friendship' do
+  	refute @self_friendship.valid?
+  end
+
+  test 'User has friend after friendship created' do
```
Comment (rvvergara) [inline test/models/friendship_test.rb]: In relation to #4 let us also then test that only one friendship is created but is reflected to both sides.

File: test/models/friend_request_test.rb [hunk 5]
```diff
@@ -0,0 +1,41 @@
+require 'test_helper'
+
+class FriendRequestTest < ActiveSupport::TestCase
```
Comment (rvvergara) [inline test/models/friend_request_test.rb]: Let us also add tests for our instance private methods `not_pending` and `not_friends`

File: test/models/post_test.rb [hunk 6]
```diff
@@ -0,0 +1,39 @@
+require 'test_helper'
+
+class PostTest < ActiveSupport::TestCase
+  def setup
+  	@post = posts(:one)
+    @user = users(:one)
+    
+  end	
+  test 'post has many comments' do
```
Comment (rvvergara) [inline test/models/post_test.rb]: Pls run me through on how we were able to create comments for our post data.

File: test/models/post_test.rb [hunk 7]
```diff
@@ -0,0 +1,39 @@
+require 'test_helper'
+
+class PostTest < ActiveSupport::TestCase
```
Comment (rvvergara) [inline test/models/post_test.rb]: Let us also add test cases for post's association with `like`

File: test/models/user_test.rb (hunk 2 above)
Comment (YellowPipe) [reply test/models/user_test.rb]: It's in the friendship fixtures, with associations with users we need

File: test/models/user_test.rb (hunk 3 above)
Comment (YellowPipe) [reply test/models/user_test.rb]: Posts fixtures

File: test/models/post_test.rb (hunk 6 above)
Comment (YellowPipe) [reply test/models/post_test.rb]: Comments fixtures =)

File: app/models/friend_request.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
+class FriendRequest < ApplicationRecord
+  belongs_to :sender, class_name: 'User'
+  belongs_to :receiver, class_name: 'User'
+ 
+  # validate :not_friends
+  validate :not_pending
+  validate :not_self
+  validate :not_friends
+
+  def accept
```
Comment (rvvergara) [inline app/models/friend_request.rb]: Great implementation here! I was going through, mentally, on the flow of how friendships will be created and then how the app will handle the associated friend_request. This answers that question in my mind! :) 

However, placing here the creation of the friendship may not be the best implementation. Later on when you do the `friendships_controller` the creation of friendship is better handled by the `create` action of that controller. That particular action will just need to invoke this method of your `friend_request` and delete the request accordingly. 

Another thing we need to consider here in the friend_request model is what happens if the receiver rejects the request (which happens lots of times on the real Facebook)? What I recommend is probably just simply calling the default `delete` method for `friend_request` for both `create` action in your `friendships_controller` and `destroy` action of your `friend_requests_controller`. The mechanisms of those will be:

1. When a receiver accepts the friend request -> Friendship record is created between the sender and the receiver -> Consequently, the friend_request record is deleted (action handled in the `friendships_controller`)

2. When a receiver rejects a friend_request -> friend_request record is deleted (action is handled in `friend_requests_controller`).

File: spec/models/post_spec.rb [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+RSpec.describe Post, type: :model do
+
+  describe 'default post details' do
+    let(:user) { create :user }
+    let(:post) { create :post }
+    let(:comment) { create :comment}
+
+
+    it 'should increase a number of posts' do
```
Comment (rvvergara) [inline spec/models/post_spec.rb]: If we're already checking for validity why do we need to check this?

File: spec/models/post_spec.rb [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+       expect(Post.all.count).to eq(post_count+1)
+    end
+
+    it 'can have many comments' do
+    	post.comments.create(body: "comment 1", user: user)
+      post.comments.create(body: "comment 2", user: user)
+      expect(post.comments.count).to eql(2)
+    end
+
+    it 'should be valid' do
```
Comment (rvvergara) [inline spec/models/post_spec.rb]: See my comment on `user_spec` on how to more properly check for validity

File: spec/models/comment_spec.rb [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+  describe 'default post details' do
+    let(:user) { create :user }
+    let(:second) { create :user }
+    let(:third) { create :user }
+    let(:post) { create :post }
+    let(:comment) { create :comment}
+
+
+    it 'should be valid at creation ' do
```
Comment (rvvergara) [inline spec/models/comment_spec.rb]: See my comment on `user_spec` for validity checking

File: app/models/friendship.rb [hunk 12]
```diff
@@ -0,0 +1,12 @@
+class Friendship < ApplicationRecord
```
Comment (rvvergara) [inline app/models/friendship.rb]: Please see #5

File: app/models/friend_request.rb [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+    destroy
+  end
+
+  private
+
+  def not_self
+    errors.add(:receiver, "can't be equal to user") if sender == receiver
+  end
+
+  def not_pending
```
Comment (rvvergara) [inline app/models/friend_request.rb]: I would rather have this check at the database level also. What if for instance there are two users in different parts of the world and they're in each other's profile? Now both of them wanna add the other and they both clicked the "Add" button at exactly the same time. These may cause failure in validating the consecutive requests. You area already making one side of this record unique in your database. You just need to make sure that the other side also happens (i.e, if user1 sends a request to user2 then at the database level we should also ensure that user2 cannot send a request to user1). 

At this moment you are carrying out the checking here at the model level. This isn't bad in itself but probably not enough to cover for the scenario that I mentioned above.

File: app/models/friend_request.rb [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+  belongs_to :sender, class_name: 'User'
+  belongs_to :receiver, class_name: 'User'
+ 
+  # validate :not_friends
+  validate :not_pending
+  validate :not_self
+  validate :not_friends
+
+  def accept
+    sender.basic_friends << receiver
```
Comment (rvvergara) [inline app/models/friend_request.rb]: We still need some more improvement here. As I created a friend request and ran this `accept` method on it I observe that the friendship that was created only reflected in the `sender` 's side but not on the `receiver` 's side. 

Here's a suggestion. You probably dont' need to have this method of accepting a friend request. What you can probably have is a callback in your friendship model that deletes any associated request whenever a friendship record is created. 

For instance I have two users `ryto` and `mike`. `ryto` sent a request to `mike`. Then when `mike` accepts the friendship then definitely this method will run in the `Frienship` model:

`Friendship.create(user_id: ryto.id, friend_id: mike.id)`

What you want is that upon the above method in `Friendship` finishing (`after_create` or `after_save`) the request between `ryto` and `mike` to be deleted.

File: spec/models/post_spec.rb [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+        post.comments.create(body: "comment 1", user: user)
+        post.comments.create(body: "comment 2", user: user)
+        expect(post.comments.count).to eql(2)
+      end
+      it 'should have any or many likes' do
+        post.likes.create(user: user)
+        expect(post.likes.count).to be >= 0 
+      end 
+
+      it 'should increase likes by one when it is liked by user' do
```
Comment (rvvergara) [inline spec/models/post_spec.rb]: Probably the test of increasing and decreasing likes count is better placed in the like_spec

File: spec/models/comment_spec.rb [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+    let(:user) { create :user }
+    let(:second) { create :user }
+    let(:third) { create :user }
+    let(:post) { create :post }
+    let(:valid_comment) { build :comment}
+    let(:comment) { create :comment}
+
+    context "Comment validations :" do
+      it 'should be valid at creation ' do
+        expect(valid_comment.valid?).to be true
```
Comment (rvvergara) [inline spec/models/comment_spec.rb]: `expect(valid_comment).to be_valid`

File: spec/models/comment_spec.rb [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+      end
+
+      it 'should be invalid without body' do
+        valid_comment.body = nil
+        expect(valid_comment.valid?).to be false
```
Comment (rvvergara) [inline spec/models/comment_spec.rb]: It's better we have a separate variable `invalid_comment` to make this more intuitive:

`expect(invalid_comment).to_not be_valid`

File: spec/models/like_spec.rb [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+RSpec.describe Like, type: :model do
+	describe 'default likes' do
+		let(:like) { build :like }
+
+		context "Like validations :" do
+			it 'should return true when like is valid' do
+		   		expect(like.valid?).to be true
+		  	end
+
+		  	it 'should return false when like is invalid becaouse of empty user' do
```
Comment (rvvergara) [inline spec/models/like_spec.rb]: Let's use another variable to make this more intuitive:
`expect(invalid_like).to_not be_valid`

File: spec/models/like_spec.rb [hunk 19]
```diff
@@ -0,0 +1,25 @@
+require 'rails_helper'
+
+RSpec.describe Like, type: :model do
```
Comment (rvvergara) [inline spec/models/like_spec.rb]: Let's rewrite some of our descriptions to make our spec readable and easily understood. See my example in `user_spec` that makes proper use of `context` to organize specs better.

File: app/models/user.rb [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+
+  has_many :comments, dependent: :destroy
+  has_many :likes, dependent: :destroy
+
+  has_many :friendships, dependent: :destroy
+  has_many :inverse_friendships, class_name: 'Friendship', foreign_key: :friend_id, dependent: :destroy
+
+  has_many :basic_friends, through: :friendships, source: :friend
+  has_many :inverse_friends, through: :inverse_friendships, source: :user
+
```
[the remark this answers is not in the document]
Comment (YellowPipe) [reply app/models/user.rb]: We talked about it on slack. I tried doing it in the models, but it was only updating friends collection after I reloaded the consoled. We agreed that we can combine it in the controller, like @friends = user.basic_friends + user.inverse_friends

File: app/models/friend_request.rb (hunk 13 above)
Comment (YellowPipe) [reply app/models/friend_request.rb]: Okay, but rails will not throw an error when the id's are opposite, how do we check this on the database level?

File: spec/models/comment_spec.rb (hunk 17 above)
Comment (YellowPipe) [reply spec/models/comment_spec.rb]: Okay, but we actually want to check if it is not valid without body, not overall. And it does not seem good to create a variable for every instance if we only use it 1 time

File: app/models/user.rb (hunk 20 above)
[the remark this answers is not in the document]
Comment (YellowPipe) [reply app/models/user.rb]: We've added friends method in the controller branch, seems to work fine.

File: app/models/friend_request.rb (hunk 13 above)
Comment (rvvergara) [reply app/models/friend_request.rb]: Excellent question @YellowPipe (and @zinagithub)!

Ok, I know you guys are awesome engineers. So all I will do is give you some clue. If you observe, in your `schema.rb` you have the following line in your `friend_requests` table:

```ruby
t.index ["receiver_id", "sender_id"], name: "index_friend_requests_on_receiver_id_and_sender_id", unique: true
```
Now, that means if we already have the ff record in the `friend_requests` table:
image

This means when we do the ff:
```
ryto = User.find(1)
jim = User.find(3)

ryto.friend_requests_sent.create(friend_id: jim.id)
```
We will get the ff:

image

The question is how to make sure the ff also gives us the same error?:

```ruby
jim.friend_requests_sent.create(receiver_id: ryto.id)
```
Ok two clues:

1. We can create a private method that will check whether there's already a friend request existing between `ryto` and `jim`. If yes, then it should add an error. You can research on how to add errors in model validation.

2. We can then have a `before_validation` callback (or `before_save`, take your pick) that will invoke this private method. This way when one of either `ryto` or `jim` sends a request to each other (when there's already an existing request between them), we will get the same error as in the picture.

**PS: You can and should also do the same for Friendship. When there's an existing friendship between two users, neither of them should be able to create a duplicate friendship with the same friend.**

File: spec/models/comment_spec.rb (hunk 17 above)
Comment (rvvergara) [reply spec/models/comment_spec.rb]: Haha you are correct @YellowPipe! Great catch! :)

File: spec/models/user_spec.rb [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+
+RSpec.describe User, type: :model do
+
+  describe 'default users' do
+    let(:valid_user) { build :user }
+    let(:user) { create :user }
+    let(:second) { create :user }
+    let(:post) { create :post }
+    
+    context 'Complete signup details' do
```
Comment (rvvergara) [inline spec/models/user_spec.rb]: I would put all validation specs under a `describe` block (within the mother describe block). I will also do the same for association specs and any instance method specs (please see my comment on `user.rb`):

```ruby
describe "User model" do
  describe "validations" do
    context "complete details" do
      .....
    end
    context "name or email is absent" do
      ....
    end
  end
  describe "associations" do
    context "user that will be deleted has posts" do
      it "also deletes all of that user's posts" do
        ..... 
     end
     context "user to be deleted has comments" do
        it "also deletes all that user's comments" do
         ....
       end  
      end
    end
   # You can write the same context block for ensuring all a user's friendships and friend requests are deleted upon user deletion
   end
   describe "instance methods" do
     # follow same pattern as above
   end
   
   describe "class methods and scopes" do
     # follow same pattern as above
   end
end
```

**Please follow this pattern in your other model specs. This makes our specs understandable and easy to follow and relate to the corresponding models being tested**

File: app/models/user.rb [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+  has_many :pending_friends, through: :friend_requests_received, source: :sender
+
+  has_many :comments, dependent: :destroy
+  has_many :likes, dependent: :destroy
+
+  has_many :friendships, dependent: :destroy
+  has_many :inverse_friendships, class_name: 'Friendship', foreign_key: :friend_id, dependent: :destroy
+
+  has_many :basic_friends, through: :friendships, source: :friend
```
[the remark this answers is not in the document]
Comment (YellowPipe) [reply app/models/user.rb]: Did that in the controller branch. I know we should have made it in this branch, but I don't want to modify it now so there are no conflicts. Point is, we have it =)

**Pull request merged**